src/librbd/AioImageRequest.h

```diff
@@ -265,12 +268,65 @@ class AioImageFlush : public AioImageRequest<ImageCtxT> {
   }
 };
 
+template <typename ImageCtxT = ImageCtx>
+class AioImageWriteSame : public AbstractAioImageWrite<ImageCtxT> {
+public:
+  AioImageWriteSame(ImageCtxT &image_ctx, AioCompletion *aio_comp, uint64_t off,
+                    uint64_t len, const char *buf, uint64_t data_len, int op_flags)
+    : AbstractAioImageWrite<ImageCtxT>(image_ctx, aio_comp, {{off, len}}),
+          m_buf(buf), m_data_len(data_len), m_op_flags(op_flags),
+          may_writesame(false) {
+    for (int i = 0; i < len / data_len; i++)
+      m_cache_bl.append(buf, data_len);
```

Comment:
Shouldn't need to allocate space if the cache isn't in use (e.g. tcmu-runner will have the cache disabled and is the primary use-case). Also, if I writesame 512bytes over 4TB of image, this is going to be very bad.